app/celery/contact_information_tasks.py

```diff
@@ -24,55 +38,197 @@
         raise NotificationPermanentFailureException(message) from e
-    except (VAProfileIDNotFoundException, VAProfileNonRetryableException) as e:
+    elif isinstance(e, (VAProfileIDNotFoundException, VAProfileNonRetryableException)):
         current_app.logger.exception(e)
         message = (
-            f'The task lookup_contact_info failed for notification {notification_id}. '
+            f'The task lookup_contact_info failed for notification {notification.id}. '
             'Notification has been updated to permanent-failure'
         )
         update_notification_status_by_id(
-            notification_id, NOTIFICATION_PERMANENT_FAILURE, status_reason=e.failure_reason
+            notification.id, NOTIFICATION_PERMANENT_FAILURE, status_reason=e.failure_reason
         )
         check_and_queue_callback_task(notification)
         raise NotificationPermanentFailureException(message) from e
+    elif isinstance(e, CommunicationItemNotFoundException):
+        current_app.logger.info(
+            'Communication item for recipient %s not found on notification %s',
+            recipient_identifier.id_value,
+            notification.id,
+        )
 
-    notification.to = recipient
-    dao_update_notification(notification)
+        return None if notification.default_send else 'No recipient opt-in found for explicit preference'
+    else:
+        current_app.logger.exception(f'Unhandled exception for notification {notification.id}: {e}')
+        raise e
+
+
+def handle_communication_not_allowed(
+    notification: Notification, recipient_identifier: RecipientIdentifier, permission_message: str
+):
+    """
+    Handles the scenario where communication is not allowed for a given notification.
+
+    Args:
+        notification (Notification): The notification object associated with the communication.
+        recipient_identifier (RecipientIdentifier): The identifier of the recipient.
+        permission_message (str): The message indicating the reason for permission denial.
+
+    Raises:
+        NotificationPermanentFailureException: If the recipient has declined permission to receive notifications.
+    """
+    if is_feature_enabled(FeatureFlag.VA_PROFILE_V3_COMBINE_CONTACT_INFO_AND_PERMISSIONS_LOOKUP):
+        current_app.logger.info(
+            'Permission denied for recipient %s for notification %s',
+            recipient_identifier.id_value,
+            notification.id,
+        )
+        reason = permission_message if permission_message is not None else 'Contact preferences set to false'
+        update_notification_status_by_id(notification.id, NOTIFICATION_PREFERENCES_DECLINED, status_reason=reason)
+
+        message = f'The recipient for notification {notification.id} has declined permission to receive notifications.'
+        current_app.logger.info(message)
+
+        check_and_queue_callback_task(notification)
+        raise NotificationPermanentFailureException(message)
```

Comment:
The next task in the celery chain is to deliver the notification.  Raising an exception here breaks the chain, preventing a notification from being sent if communication preferences prevent it.

Reply:
It occurs to me that the message here does not differentiate between whether a user explicitly declined to receive the communication, or if a determination not to send the notification was made due to the `default_send` behavior.  

This information is present in the `reason` variable used to update the notification status, but not in the `message` which is passed to the logger/raised in exception.